mtd spi-nor: use 4K erase only where need, with 64K erasesize otherwise

john-tho opened the pull request:
This PR allows a 4K erasesize to be used only where needed and supported by the underlying SPI-NOR chip, while otherwise using a 64K erasesize.

This makes large writes (example: flashing a sysupgrade image) on spi-nor that was using `USE_4K_SECTORS` significantly faster, as there are fewer erase commands that the NOR chip must complete.

Required kernel config:
- `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS=n`
  `y` forces 4K sector size to be used globally, and must be disabled to allow 64K erasesize.
- `CONFIG_MTD_SPI_NOR_USE_VARIABLE_ERASE=y`
  introduced by this PR, and must be set to allow 4K erases to be selectively used.

The parent of a writeable partition must also be writeable.
See [partitions.h mtd_partition mask_flags](https://elixir.bootlin.com/linux/v5.4.80/source/include/linux/mtd/partitions.h#L36), [mtdpart.c child uses masked parent flags](https://elixir.bootlin.com/linux/v5.4.50/source/drivers/mtd/mtdpart.c#L344)

There are two know situations where a 4K erase sector is required in Openwrt:
- modification of RedBoot FIS directory flash partition layout, on some devices
  [This should be modified to use the partition size, rather than the erase size](https://github.com/openwrt/openwrt/blob/d643b10a76965a94684012d2ddf36e2bcc146274/package/system/mtd/src/fis.c#L79)
- modification of Mitrotik RouterBoot soft_config bootloader configuration
  [This should be modified to work with `USE_VARIABLE_ERASE`](https://github.com/openwrt/openwrt/blob/a4c1c169c391ac0b11e4c3e895b282f07585347c/target/linux/generic/files/drivers/platform/mikrotik/rb_softconfig.c#L67)

Prior to kernel 4.19, [`470-mtd-spi-nor-support-limiting-4K-sectors-support-base.patch`](https://github.com/openwrt/openwrt/blob/master/target/linux/generic/pending-5.4/470-mtd-spi-nor-support-limiting-4K-sectors-support-base.patch) worked with  `USE_4K_SECTORS`, and allowed 4K sectors to be used on partitions smaller than a configured size, while larger partitions used 64K erasesize. This patch is non-functional on 5.4 due to kernel changes.

The current workaround, [`411-mtd-partial_eraseblock_write.patch`](https://github.com/openwrt/openwrt/blob/master/target/linux/generic/pending-5.4/411-mtd-partial_eraseblock_write.patch) & [`412-mtd-partial_eraseblock_unlock.patch`](https://github.com/openwrt/openwrt/blob/master/target/linux/generic/pending-5.4/412-mtd-partial_eraseblock_unlock.patch), allows fake 4K erasesize by coping 64K, erasing the block, then restoring the remainder not needed for the 4K erase. There are no safety checks through this, and [a merge bug in 4.19 & 5.4](https://github.com/openwrt/openwrt/commit/bf2870c1d9e162f27925e9c40eca9e0069096f3b) was causing data loss and unbootable devices. Kernel changes in 5.7 will require this fake 4K erase patch to be reworked or removed.

PR opened in response to seeing data loss, and slow NOR erase: https://github.com/openwrt/openwrt/pull/3103#issuecomment-667610510

The commit `testing-only: kernel no-op spi-nor write & erase` is testing-only, and needs to be removed from this PR to allow erase and write.

adrianschmutzler commented:
@dwmw2 Maybe you can give feedback/help on this one?

john-tho commented:
Yes, I missed things in the multi_sector_erase path, but it showed that my changed 472-mtd-spi-nor-attempt-4K-erase.patch now writes to the correct location.

```
root@OpenWrt:~# dmesg
[    5.012627] Creating 2 MTD partitions on "spi0.0":                           
[    5.017419] 0x000000000000-0x000000040000 : "RouterBoot"                     
[    5.025040] 5 routerbootpart partitions found on MTD device RouterBoot
[    5.031575] Creating 5 MTD partitions on "RouterBoot":                      
[    5.036708] 0x000000000000-0x00000000f000 : "bootloader1"
[    5.043398] 0x00000000f000-0x000000010000 : "hard_config"                   
[    5.050025] 0x000000010000-0x00000001f000 : "bootloader2"                  
[    5.056696] 0x000000020000-0x000000021000 : "soft_config"                    
[    5.063404] 0x000000030000-0x000000031000 : "bios" 

root@OpenWrt:~# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00040000 00010000 "RouterBoot"
mtd1: 0000f000 00010000 "bootloader1"
mtd2: 00001000 00010000 "hard_config"
mtd3: 0000f000 00010000 "bootloader2"
mtd4: 00001000 00001000 "soft_config"
mtd5: 00001000 00010000 "bios"
mtd6: 00fc0000 00010000 "firmware"

# the 4K erasesize on mtd4 soft_config comes from
# target/linux/generic/pending-5.4/472-mtd-spi-nor-attempt-4K-erase.patch
# and not having CONFIG_MTD_SPI_NOR_USE_4K_SECTORS defined

# test mtdpart.c
# added mtd_debug to utils/mtd-utils/nand-utils

root@OpenWrt:~# /tmp/usr/sbin/mtd_debug erase /dev/mtd4 0x0 4096
Erased 4096 bytes from address 0x00000000 in flash
root@OpenWrt:~# dmesg | tail
[ 3715.284612] spi-nor spi0.0: at 0x20000, len 4096
[ 3715.289272] spi-nor spi0.0: spi_nor_erase uniform erase too big for erasesize, try multi_sector erase.
[ 3715.298593] mtdpart_erase initial attempt failed: -22
[ 3715.303649] mtdpart_erase overflow manual copy,erase,write: soft_config
# mtd_read
[ 3715.310291] spi-nor spi0.0: from 0x00020000, len 65536
# mtd_erase
[ 3715.346876] spi-nor spi0.0: at 0x20000, len 65536
[ 3715.351606] spi-nor spi0.0: erase_sector on spi0.0 erase_address: 0x20000 erase_opcode: 0xd8
[ 3715.360135] spi-nor spi0.0: nulled erase_sector
# mtd_write
[ 3715.364728] spi-nor spi0.0: to 0x00021000, len 61440
[ 3715.369702] spi-nor spi0.0: nulled spi_nor_write
```

john-tho commented:
I might be seeing an issue with regions = &uniform_region?

https://elixir.bootlin.com/linux/v5.4.51/source/include/linux/mtd/spi-nor.h#L329
https://elixir.bootlin.com/linux/v5.4.51/source/drivers/mtd/spi-nor/spi-nor.c#L3308
```
struct spi_nor_erase_map {
	struct spi_nor_erase_region	*regions;
	struct spi_nor_erase_region	uniform_region;
	struct spi_nor_erase_type	erase_type[SNOR_ERASE_TYPE_MAX];
	u8				uniform_erase_type;
};
static void spi_nor_init_uniform_erase_map(struct spi_nor_erase_map *map,
					   u8 erase_mask, u64 flash_size)
{
	/* Offset 0 with erase_mask and SNOR_LAST_REGION bit set */
	map->uniform_region.offset = (erase_mask & SNOR_ERASE_TYPE_MASK) |
				     SNOR_LAST_REGION;
	map->uniform_region.size = flash_size;
	map->regions = &map->uniform_region;
	map->uniform_erase_type = erase_mask;
}
```

The multi_sector path uses regions, but I am seeing the wrong data (should be == uniform_region)?
```
[    5.035231] init_uniform_erase_map                                
[    5.038677] erase_mask: 0x3 flash_size: 0x1000000 SNOR_LAST_REGION: 0x10 SNOR_ERASE_TYPE_MASK: 0xf
[    5.047628] uniform_region: 0x(ptrval) offset: 0x13 size: 0x1000000
[    5.053886] regions: 0x(ptrval) offset: 0x13 size: 0x1000000
[    5.054009] init_uniform_erase_map                                 
[    5.063139] erase_mask: 0x7 flash_size: 0x1000000 SNOR_LAST_REGION: 0x10 SNOR_ERASE_TYPE_MASK: 0xf
[    5.072081] uniform_region: 0x(ptrval) offset: 0x17 size: 0x1000000
[    5.078331] regions: 0x(ptrval) offset: 0x17 size: 0x1000000
[  395.366758] spi-nor spi0.0: spi_nor_init_erase_cmd_list
[  395.372016] find_erase_region addr: 0x20000 region pointer 0xd5ef35e5 region->offset: 0x8ff1dfa081021e40 region->size: 0x8ff1e3ec00000000
[  395.384528] find_erase_region uniform_region uniform_region pointer 0x6d111ff5 uniform_region->offset: 0x1e uniform_region->size: 0x1000000
```

https://elixir.bootlin.com/linux/v5.4.51/source/drivers/mtd/spi-nor/spi-nor.c#L1057


The partial erase fix sent to openwrt-devel for feedback

f00b4r0 commented:
Maybe it would be worth getting linux-mtd involved for review/feedback?

At the end of the day what we need is for a working solution that is fully upstreamed: what killed the current patch set is the fact that it never got into mainline, it's probably wise to avoid this pitfall now.

john-tho commented:
I will go to linux-mtd about allowing uniform erase region devices to use the multi_sector path for a smaller erase (472-0-mtd-spi-nor-attempt-4K-erase.patch), less the mtdpart.c changes.

Seems to be working:
```
mtd_debug erase /dev/mtd4 0x000 4096
[  146.383922] spi-nor spi0.0: at 0x20000, len 4096
[  146.388715] spi-nor spi0.0: erase_sector on spi0.0 erase_address: 0x20000 erase_opcode: 0x20 
[  146.397254] spi-nor spi0.0: no-op erase_sector

mtd_debug erase /dev/mtd0 0x20000 65536
[  180.140322] spi-nor spi0.0: at 0x20000, len 65536
[  180.145173] spi-nor spi0.0: erase_sector on spi0.0 erase_address: 0x20000 erase_opcode: 0xd8 
[  180.153706] spi-nor spi0.0: no-op erase_sector

mtd_debug erase /dev/mtd0 0x20000 69632
[  189.157406] spi-nor spi0.0: at 0x20000, len 69632
[  189.162217] spi-nor spi0.0: erase_sector on spi0.0 erase_address: 0x20000 erase_opcode: 0xd8 
[  189.170773] spi-nor spi0.0: no-op erase_sector
[  189.175272] spi-nor spi0.0: erase_sector on spi0.0 erase_address: 0x30000 erase_opcode: 0x20 
[  189.183804] spi-nor spi0.0: no-op erase_sector

mtd_debug erase /dev/mtd0 0x20000 65536
[  225.025982] spi-nor spi0.0: at 0x20000, len 65536
[  225.030802] spi-nor spi0.0: erase_sector on spi0.0 erase_address: 0x20000 erase_opcode: 0xd8 
[  225.039350] spi-nor spi0.0: no-op erase_sector
```

john-tho commented on target/linux/generic/pending-5.4/411-mtd-partial_eraseblock_write.patch:
```diff
@@ -73,10 +73,9 @@ Signed-off-by: Felix Fietkau <nbd@nbd.name>
  	instr->addr += part->offset;
  	ret = part->parent->_erase(part->parent, instr);
-@@ -213,6 +262,24 @@ static int part_erase(struct mtd_info *m
+ 	if (instr->fail_addr != MTD_FAIL_ADDR_UNKNOWN)
  		instr->fail_addr -= part->offset;
```
I should change this, so that if the erase fails before the write address, we avoid the write?

john-tho commented on target/linux/generic/pending-5.4/472-0-mtd-spi-nor-attempt-4K-erase.patch:
```diff
@@ -0,0 +1,159 @@
+ 	if (!erase)
+ 		return NULL;
+ 
+-	/* Disable all other Sector Erase commands. */
+-	map->uniform_erase_type &= ~SNOR_ERASE_TYPE_MASK;
+-	map->uniform_erase_type |= BIT(erase - map->erase_type);
++	/* Disable all other Sector Erase commands.
++	 * map->uniform_erase_type &= ~SNOR_ERASE_TYPE_MASK;
++	 * map->uniform_erase_type |= BIT(erase - map->erase_type);
```
I have not looked into if anything else (the sector-at-a-time part) uses `map->uniform_erase_type` for uniform erase region spi-nor.

john-tho committed:
mtd-utils: add mtd-utils-debug

Provides easy access to low level mtd operations:

mtd_debug info <device>
mtd_debug read <device> <offset> <len> <dest-filename>
mtd_debug write <device> <offset> <len> <source-filename>
mtd_debug erase <device> <offset> <len>

http://git.infradead.org/mtd-utils.git/blob/HEAD:/misc-utils/mtd_debug.c

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>

john-tho commented:
I have just force-pushed a new version I am much happier with.

Still open to suggestions.

A kernel config variable `CONFIG_MTD_SPI_NOR_USE_VARIABLE_ERASE=y` is introduced that can disable the standard uniform erase regions code path, forcing the multiple erase regions path to be taken. This multiple erase regions path will use 4K erase operations where needed and supported.

An `erasesize_minor` variable is added to the `struct mtd_info`, so that the mtd level does not need to understand the spi-nor level to calculate the smallest (_minor_) universal erasesize. This is set in the spi-nor device initialisation. Here, I am not yet certain the erasesize that is picked is universal.

mtdpart is modified so that it will not prevent writing to a partition that starts or ends on a _minor_ erasesize boundary.

In `spi_nor_sfdp_init_params`, without (re)setting the `erase_map.regions` = `&erase_map.uniform_region` after the memcpy, the region pointer is bogus. I think this is a bug upstream?

Is [`redboot_fis_do_upgrade`](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/generic/base-files/lib/upgrade/platform.sh) the other situation (Redboot), where partial erases were used? I have not worked out that code, or the intention of it yet? If is does not to a modulo 4K erase / write, it would need to be changed to retain functionality. One suggestion from openwrt-devel was doing this in userspace, with something like: https://github.com/jclehner/mtd-rw.

Our [Openwrt Mikrotik soft_config write usage](https://github.com/openwrt/openwrt/blob/master/target/linux/generic/files/drivers/platform/mikrotik/rb_softconfig.c) does write the whole 4K, so it should work fine with this (after removing the `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS` test).

I have only done testing with my no-oped writes / erases so far, but seeing what I am expecting:
```
[    5.178142] Creating 2 MTD partitions on "spi0.1": 
[    5.183013] 0x000000000000-0x000000020000 : "RouterBoot"
[    5.190780] 5 routerbootpart partitions found on MTD device RouterBoot
[    5.197444] Creating 5 MTD partitions on "RouterBoot":
[    5.202670] 0x000000000000-0x00000000e000 : "bootloader1"
[    5.208931] 0x00000000e000-0x00000000f000 : "hard_config"
[    5.215205] 0x00000000f000-0x000000010000 : "bios"
[    5.220873] 0x000000010000-0x00000001f000 : "bootloader2"
[    5.227157] 0x00000001f000-0x000000020000 : "soft_config"
[    5.243235] 0x000000020000-0x000001000000 : "firmware"

root@OpenWrt:~# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00020000 00010000 "RouterBoot"
mtd1: 0000e000 00010000 "bootloader1"
mtd2: 00001000 00010000 "hard_config"
mtd3: 00001000 00010000 "bios"
mtd4: 0000f000 00010000 "bootloader2"
mtd5: 00001000 00001000 "soft_config"
mtd6: 00fe0000 00010000 "firmware"
# the erasesize here is only for show, it is not used

# 4K erase
root@OpenWrt:~# mtd_debug erase /dev/mtd5 0x000 4096
Erased 4096 bytes from address 0x00000000 in flash
[ 2521.775540] spi-nor spi0.1: at 0x1f000, len 4096
[ 2521.775681] spi-nor spi0.1: erase_sector on spi0.1 erase_address: 0x1f000 erase_opcode: 0x20 

# erase that overflows partition
root@OpenWrt:~# mtd_debug erase /dev/mtd5 0x000 65536
MEMERASE: Invalid argument

# 64K erase
root@OpenWrt:~# mtd_debug erase /dev/mtd0 0x0000 65536
Erased 65536 bytes from address 0x00000000 in flash
[ 2629.396167] spi-nor spi0.1: at 0x0, len 65536
[ 2629.396308] spi-nor spi0.1: erase_sector on spi0.1 erase_address: 0x0 erase_opcode: 0xd8 

# 64K+4K erase
root@OpenWrt:~# mtd_debug erase /dev/mtd0 0x0000 69632                          
Erased 69632 bytes from address 0x00000000 in flash 
[ 2748.714409] spi-nor spi0.1: at 0x0, len 69632
[ 2748.714543] spi-nor spi0.1: erase_sector on spi0.1 erase_address: 0x0 erase_opcode: 0xd8 
[ 2748.728218] spi-nor spi0.1: erase_sector on spi0.1 erase_address: 0x10000 erase_opcode: 0x20 
```

Cheers!

john-tho commented:
Hi @tmn505, could you please help me out with the RedBoot sysupgrade code?
I have been skipping over the code, and https://github.com/openwrt/openwrt/pull/1237.
I am still pretty unsure, but it looks to me like:
For RedBoot:
The flash partition scheme is defined on-flash, in the "FIS directory".
This is parsed by the [kernel redboot parser](https://elixir.bootlin.com/linux/v5.4.66/source/drivers/mtd/parsers/redboot.c), which uses the parent mtd erasesize in parsing, but accounts for _FIS directory_ and _RedBoot config_ combined on one eraseblock. This parser generates `parts`, an array of `struct mtd_partition`, setting name, offset, and size for each. This is used by `mtdpart` through `pparts`. Guessing this eventually goes through `mtdpart` `allocate_partition`, where I sometimes set and erasesize differently.

[jjPlus JA76PF2 RedBoot partitions example](https://bugs.openwrt.org/index.php?do=details&task_id=2428)
This was made with the 411 partial_erase patch, so the erasesize was set to partition size, where size<eraseblock.
```
dev:    size   erasesize  name
mtd0: 00040000 00010000 "RedBoot"
mtd1: 00120000 00010000 "linux"
mtd2: 00e80000 00010000 "rootfs"
mtd3: 00c60000 00010000 "rootfs_data"
mtd4: 0000f000 0000f000 "FIS directory"
mtd5: 00001000 00001000 "RedBoot config"
```
On sysupgrade, [redboot_fis_do_upgrade](https://github.com/openwrt/openwrt/blob/d643b10a76965a94684012d2ddf36e2bcc146274/target/linux/ath79/generic/base-files/lib/upgrade/platform.sh#L8) uses [Openwrt's mtd](https://github.com/openwrt/openwrt/blob/master/package/system/mtd/src/mtd.c) to write the combined kernel and rootfs, and also update the FIS data on flash.
I think mtd does use the partition erasesize for partitions: https://github.com/openwrt/openwrt/blob/90853439a162d7c5389f6a68028b4390c349c04d/package/system/mtd/src/mtd.c#L143

I am not sure if the FIS remap uses the _FIS directory_ erasesize (I don't understand the `external erasesize` and program flow): https://github.com/openwrt/openwrt/blob/d643b10a76965a94684012d2ddf36e2bcc146274/package/system/mtd/src/fis.c#L79 This may need to be changed to use the size of the _FIS directory_ partition instead, if my patch sets the _FIS directory_ erasesize to 1K?

Any feedback would be great.
Cheers

john-tho commented:
Any suggestions for further review?
I am thinking I send the `spi_nor_sfdp_init_params`, `if (erase_map.regions = &erase_map.uniform_region)`, re-set after memcpy patch alone to linux-mtd, then if I get any action on that, follow up with the `mtdpart allocate_partition` allow writable minor boundary changes?

Send my whole `allow use of 4K erase` commit, with a note to cut 411, 412, and 470, to openwrt-devel?

Thanks

f00b4r0 commented:
Hi,

Thanks for keeping this alive. I can't really comment on the merit of your new version over the previous one, I'm just not familiar enough with this code. However what I would strongly advocate is for you to try and get this pushed upstream (linux-mtd): we really need to get as much of this upstreamed as possible, so as to make sure future mtd API changes don't stab us in the back going forward.

Thus I think you should send this patch to both openwrt-devel and linux-mtd. Maybe @dwmw2 would be able to suggest who to ping then to get some feedback on this?

Regarding the option to use mtd-rw, I think it's just a bad idea. It entirely defeats the purpose of marking some partitions read-only and will make it that much easier for users to brick their devices. I'd avoid it at all costs. Furthermore, adding a userspace layer to process mtd writes (if I understand the intent correctly) also seems like a Very Bad Idea™, especially in the context of low-memory devices.

My 2c,
Thibaut

tmn505 commented:
> Hi @tmn505, could you please help me out with the RedBoot sysupgrade code?
> I have been skipping over the code, and #1237.
> I am still pretty unsure, but it looks to me like:
> For RedBoot:
> The flash partition scheme is defined on-flash, in the "FIS directory".

Yes.

> This is parsed by the [kernel redboot parser](https://elixir.bootlin.com/linux/v5.4.66/source/drivers/mtd/parsers/redboot.c), which uses the parent mtd erasesize in parsing, but accounts for _FIS directory_ and _RedBoot config_ combined on one eraseblock.

Yes.

> This parser generates `parts`, an array of `struct mtd_partition`, setting name, offset, and size for each. This is used by `mtdpart` through `pparts`. Guessing this eventually goes through `mtdpart` `allocate_partition`, where I sometimes set and erasesize differently.

I think yes, would have much more to say if I understood C.

> 
> [jjPlus JA76PF2 RedBoot partitions example](https://bugs.openwrt.org/index.php?do=details&task_id=2428)
> This was made with the 411 partial_erase patch, so the erasesize was set to partition size, where size<eraseblock.
> 
> ```
> dev:    size   erasesize  name
> mtd0: 00040000 00010000 "RedBoot"
> mtd1: 00120000 00010000 "linux"
> mtd2: 00e80000 00010000 "rootfs"
> mtd3: 00c60000 00010000 "rootfs_data"
> mtd4: 0000f000 0000f000 "FIS directory"
> mtd5: 00001000 00001000 "RedBoot config"
> ```
> 
> On sysupgrade, [redboot_fis_do_upgrade](https://github.com/openwrt/openwrt/blob/d643b10a76965a94684012d2ddf36e2bcc146274/target/linux/ath79/generic/base-files/lib/upgrade/platform.sh#L8) uses [Openwrt's mtd](https://github.com/openwrt/openwrt/blob/master/package/system/mtd/src/mtd.c) to write the combined kernel and rootfs, and also update the FIS data on flash.
> I think mtd does use the partition erasesize for partitions:
> 
> https://github.com/openwrt/openwrt/blob/90853439a162d7c5389f6a68028b4390c349c04d/package/system/mtd/src/mtd.c#L143
> 
> I am not sure if the FIS remap uses the _FIS directory_ erasesize (I don't understand the `external erasesize` and program flow):
> 
> https://github.com/openwrt/openwrt/blob/d643b10a76965a94684012d2ddf36e2bcc146274/package/system/mtd/src/fis.c#L79

I think mtd is happy with whatever the partition size and eraseblock size is, since the part witting FIS directory was written with partial eraseblock patches in place. There are cases where FIS directory and RedBoot config are on separate eraseblock, and it works there to.

> This may need to be changed to use the size of the _FIS directory_ partition instead, if my patch sets the _FIS directory_ erasesize to 1K?

That's FIS directory in case of the devices in my possession but that can differ (not between the same models but different products). I don't know how far would be the best case.

john-tho commented:
Hi @robimarko,

This is the branch I have been using to test my 4K changes.

Needs kernel config:
```
CONFIG_MTD_SPI_NOR_USE_VARIABLE_ERASE=y
CONFIG_MTD_SPI_NOR_USE_4K_SECTORS=n
```

I started with mtd-debug, the no-op commit, and dynamic_debug for testing
```
CONFIG_PACKAGE_mtd-utils-debug=y
CONFIG_KERNEL_DYNAMIC_DEBUG=y
```
`echo -n 'file spi-nor.c +p' > /sys/kernel/debug/dynamic_debug/control`, so the erases show on dmesg

`mtd_debug erase $(soft_config_partition_dev) 0x0 4096` should work
`mtd_debug erase $(soft_config_partition_dev) 0x0 4095` should fail: 4095 not a full eraseblock
`mtd_debug erase $(soft_config_partition_dev) 0x0 65536` should fail: over partition length
`mtd_debug erase $(firmware_partition_dev) 0x0 65536` should work
`mtd_debug erase $(firmware_partition_dev) 0x0 $(( 2 ** 16 + 2 ** 12))` should work: 2 64K blocks + 2 4K blocks.

If that works, change rb_softconfig to allow writes:
https://github.com/openwrt/openwrt/blob/0830ae3a2fdf444380153d311a7de167b92a68af/target/linux/generic/files/drivers/platform/mikrotik/rb_softconfig.c#L67

Then revert the no-op patch.

A cleaner patch to fix the spi-nor memcpy got reviewed-by a Maintainer on linux-mtd:
https://lists.infradead.org/pipermail/linux-mtd/2020-October/082821.html
I will add that patch after I test it.

champtar commented:
@john-tho I can test that on my `MikroTik RouterBOARD wAP G-5HacT2HnD` when it's "safe" & ready to test, just ping me
(and tell me what to look for ;) )

john-tho commented:
Thanks,
What we really need is someone to go through my kernel changes, and okay what I have done: https://github.com/openwrt/openwrt/pull/3271/commits/b27818cefe6a82d4045427be1cbda43f65553fa8
I guess I can send just this through to linux-mtd as RFC, but I have had no responses on an earlier attempt (it was a much dirtier approach), and a bug fix (a different fix for the same bug was reviewed later).

f00b4r0 commented:
@john-tho I would ping again on linux-mtd and openwrt-devel. Eventually you will get an answer. No answer isn't an answer.

f00b4r0 commented:
By the way I would probably remove the "Draft" status on this PR, so it gets a bit more attention.

adschm commented:
Being ready-for-review, you should also add commit messages and Signed-off-by to all patches.

john-tho commented:
Thanks @adschm.
Pushed with commit messages and signoff.
Cut spi-nor no-op patch.

Now needs more review / testing, and [an overview, or fix for RedBoot FIS rewrite before merge](https://github.com/openwrt/openwrt/blob/d643b10a76965a94684012d2ddf36e2bcc146274/package/system/mtd/src/fis.c#L79): Should FIS rewrite use partition size, rather than erasesize?

Cheers

adschm commented:
> Now needs more review / testing, and an overview, or fix for RedBoot FIS rewrite before merge: Should FIS rewrite use partition size, rather than erasesize?

@tmn505 Can you comment on this?

adschm commented:
"package: add mtd-utils-debug" -> "mtd-utils: add mtd-utils-debug"

john-tho committed:
ipq40xx: mikrotik: define common NOR layout

All Mikrotik ipq40xx devices use a common NOR partition layout
Reduce DTS duplication by using a DTSI

Limit the RouterBoot partition to length 0x40000, as done by OEM:
	0x000000080000-0x0000000c0000 : "RouterBoot"
Enable m25p,fast-read
Set RouterBoot parent partition as rw,
so that the child soft_config partition can be made writable

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>

john-tho commented:
Hi @danielg4, @ynezz, @nbd168 I saw you had commited to Openwrt mtd FIS or @tmn505:
Could you let me know if an FIS directory rewrite should always read and rewrite the whole FIS directory partition?
If so, could we change these:
[`desc = mmap(NULL, erasesize, PROT_READ|PROT_WRITE, MAP_SHARED|MAP_LOCKED, fis_fd, 0);`](https://github.com/openwrt/openwrt/blob/f281aca97821ece3990b3082bfb14b6fa63ddec4/package/system/mtd/src/fis.c#L80)
[`msync(fis_desc, fis_erasesize, MS_SYNC|MS_INVALIDATE);`](https://github.com/openwrt/openwrt/blob/f281aca97821ece3990b3082bfb14b6fa63ddec4/package/system/mtd/src/fis.c#L238)

The `411-mtd-partial_eraseblock_write.patch` patch set partition erasesize to partition size.
In my replacement here, where needed, partition erasesize is set to the smallest erasesize supported by the flash.

This example would break Openwrt `mtd` `fis.c` if the erasesize is 0x1000?
[jjPlus JA76PF2 RedBoot partitions example](https://bugs.openwrt.org/index.php?do=details&task_id=2428)
```
dev:    size   erasesize  name
mtd4: 0000f000 0000f000 "FIS directory"
```

Also, is an FIS directory partition size is expected to be divisible by 0x1000?
If not, there will still need to be some sort of partial erase support.

Cheers

danielg4 commented:
Partial erase support is a must, because multiple partitions per erase block are just too common. Due to changes in the kernel with regard to writing to part of a block without overwriting the rest of it, my understanding is that the whole erase block must be written at once from userspace now.

john-tho commented:
@danielg4
> multiple partitions per erase block are just too common.

If you know of cases where partitions are not `0x1000` aligned, please let me know.

> the whole erase block must be written at once from userspace now.

Yes, I think non-erasesize operations should be done in userspace, but there is currently a patch that does this in Openwrt kernel: https://github.com/openwrt/openwrt/blob/master/target/linux/generic/pending-5.4/411-mtd-partial_eraseblock_write.patch. This needs to change with kernel v5.7.
Most current flash chips supports a 4K erase, but this is generally not exposed unless 4K is forced with `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS=y`.

No comment on FIS Directory rewriting the whole FIS directory partition?

Cheers

danielg4 commented:
The whole FIS directory partition should be written like any other erase block, since it never uses more than one. Any partition sharing the erase block with it should be written at the same time, of course. Look at the BlueSocket devices for tiny partitions sharing erase blocks. Does ```CONFIG_MTD_SPI_NOR_USE_4K_SECTORS``` check for chips that don't support it?

f00b4r0 commented:
> @danielg4
> 
> > multiple partitions per erase block are just too common.
> 
> If you know of cases where partitions are not `0x1000` aligned, please let me know.

It should actually be "writeable partitions" I think. We don't care about read-only partition since by definition we will never erase them.

> > the whole erase block must be written at once from userspace now.
> 
> Yes, I think non-erasesize operations should be done in userspace,

I personally believe that any attempt to address this situation in userspace is an open door to bricking devices. In low memory situations, the probability of failure is high.

Reading the fis code, it seems the intent really is to mmap the whole partition, in which case the code should be adjusted. The kernel mtd layer should handle multiple (small) EBs from a larger mmap, AIUI.

My 2c.

f00b4r0 commented:
@john-tho @adschm tells me we're very close to the next OpenWRT release, did you get any further with this upstream?

john-tho commented:
I have not sent it upstream yet. Have [updated and cleaned up a bit more for 5.10][1], and waiting to see any of the maintainers active on linux-mtd again.

[1]: https://github.com/john-tho/openwrt/blob/ramips_5.10_hexs/target/linux/generic/pending-5.10/499-0-mtd-spi-nor-allow-variable-erase.patch

Leo-PL commented:
@john-tho Any update on upstreaming?
I just ported OpenWrt to ZTE MF283+ which has 16MB Flash capable of sector erase, and probably due to lack of this patch, the performance is really, really bad, even if I bump clock to 50MHz and enable fast reads. I'll test it for you in a while, too.

john-tho commented:
No response upstream on linux-mtd so far (submitted a month ago). I sent a few more queries today. https://lore.kernel.org/linux-mtd/20210104122853.18428-1-git@johnthomson.fastmail.com.au/T/#u

Any testing or review is welcome, thanks.

Leo-PL commented:
I'll test it soon, especially because I noticed, that on OpenWrt master, option to limit use of 4K sectors to below defined value failed me today, so there is a bug lurking in OpenWrt patches somewhere. I had to disable 4K sectors entirerly to get performance gain - I'll bring it up with OpenWrt team separately though :-)

Leo-PL commented:
It took a while, but I had a chance of testing this on rt305x platform, on my ZTE MF283+ which would be very worthy target for this fix. However, I stumbled on the issue that partitions were forced read-only due to not being on erase block boundary. Could you rebase your patch against recent kernel?

Boot log is attached:
[bootlog_mtd_variable_erase.txt](https://github.com/openwrt/openwrt/files/6368266/bootlog_mtd_variable_erase.txt)

john-tho commented:
Hi @Leo-PL. Yes, I need to rebase. Thanks.

Partitions being read-only if they do not start / finish on an eraseblock boundary is expected. It should not matter (_kernel_ does not change, and _rootfs_ is read-only squashfs). Sysupgrade should use the parent _firmware_ partition.

>```
>[    0.706265] 0x000000080000-0x000001000000 : "firmware"
>[    0.726455] 2 uimage-fw partitions found on MTD device firmware
>[    0.738309] Creating 2 MTD partitions on "firmware":
>[    0.748220] 0x000000000000-0x0000001b5616 : "kernel"
>[    0.758105] mtd: partition "kernel" doesn't end on an erase/write block -- force read-only
>[    0.776398] 0x0000001b5616-0x000000f80000 : "rootfs"
>[    0.786353] mtd: partition "rootfs" doesn't start on an erase/write block boundary -- force read-only
>[    0.806448] mtd: device 5 (rootfs) set to be root filesystem
>[    0.822004] 1 squashfs-split partitions found on MTD device rootfs
>[    0.834398] 0x0000006b1000-0x000000f80000 : "rootfs_data"
>```

If you do not want them read-only, pad them in your image recipes.
Something like (needs to change to suit device):
```
BLOCKSIZE := 64k
IMAGE/sysupgrade.bin := append-kernel | pad-to $$$$(BLOCKSIZE) |\
	append-rootfs | pad-rootfs | append-metadata | check-size
```

f00b4r0 commented:
There shouldn't be any reason to have the kernel and rootfs partition RW.

Leo-PL commented:
> 
> 
> There shouldn't be any reason to have the kernel and rootfs partition RW.

Yes, but rootfs_data resulting from squashfs-split should be RW, without that system fails to boot. I'll try again with BLOCKSIZE set to 64k.

Leo-PL commented:
Ok, I did one more test - the system worked, it just took 5 minutes to format rootfs_data partition after reboot. Maybe I missed something with configuration, but I don't see improvement in performance of sysupgrade - it takes around 8 to 10 minutes on this device, just like it did without patch, of course - the time varies a bit.

What's strange, is that I still see 4K erase blocks even if I lower CONFIG_MTD_SPI_NOR_USE_4K_SECTORS_LIMIT below 16M. It only does when I disable CONFIG_MTD_SPI_NOR_USE_4K_SECTORS entirely. I think that `mtd` still uses smallest possible erase size, as seen in `/proc/mtd`
With CONFIG_MTD_SPI_NOR_USE_4K_SECTORS disabled, `sysupgrade` takes around 2:30 on this device.
I'll raise this issue on openwrt-devel.

Leo-PL commented:
I did a proper test (finally) and the performance is improved, however I suspect that's simply due to the fact, that JFFS2 simply uses 64k-sized erase blocks unconditionally, and I could get the same effect by just disabling CONFIG_MTD_SPI_NOR_USE_4K_SECTORS, but with no benefits in term of space availability, which matters for "tiny" devices.a

I see that for proper test, I should use a different device, which has smaller partitions. For example, introduce a writable U-boot environment on ath79 device equipped with u-boot_mod, which would be stored as single pages in an erase block used by U-boot itself. And then, the kernel should allow me to write to it.
Besides, I noticed that migrating between 4k and 64k-sectored JFFS2 over `sysupgrade` just works, aside from a few warnings in the kernel, as filesystem on rootfs_data is created from scratch on sysupgrade anyway.

Leo-PL commented:
I performed aforementioned test.
MTD configuration  is:
```
[    0.377319] 4 fixed-partitions partitions found on MTD device spi0.0
[    0.383785] Creating 4 MTD partitions on "spi0.0":
[    0.388667] 0x000000000000-0x000000020000 : "u-boot"
[    0.394763] 0x00000001ec00-0x00000001fc00 : "u-boot-env"
[    0.400223] mtd: partition "u-boot-env" doesn't start on an erase/write block boundary -- force read-only
[    0.410984] 0x000000020000-0x0000007f0000 : "firmware"
[    0.420082] 2 tplink-fw partitions found on MTD device firmware
[    0.426149] Creating 2 MTD partitions on "firmware":
[    0.431231] 0x000000000000-0x0000001d9c23 : "kernel"
[    0.436289] mtd: partition "kernel" doesn't end on an erase/write block -- force read-only
[    0.445699] 0x0000001d9c24-0x0000007d0000 : "rootfs"
[    0.450805] mtd: partition "rootfs" doesn't start on an erase/write block boundary -- force read-only
[    0.461151] mtd: device 4 (rootfs) set to be root filesystem
[    0.468875] 1 squashfs-split partitions found on MTD device rootfs
[    0.475219] 0x000000720000-0x0000007d0000 : "rootfs_data"
[    0.481763] 0x0000007f0000-0x000000800000 : "art"
```
`/etc/fw_env.config` is:
```
/dev/mtd1 0x0000 0x1000 0x1000
```
`fw_printenv` works, but fw_setenv fails with `EPERM`, just as it would without the patch.

When I alter `/etc/fw_env.config` to use U-boot partition:
```
/dev/mtd0 0x1EC00 0x1000 0x1000
```
`fw_printenv` complains about environment not starting on erase block boundary - I'm not sure if it is expected with the patch, or not.

john-tho commented:
I have pushed an updated patch for SPI-NOR variable erase support.
This is based on what I sent to lkml & linux-mtd (no response there)

One patch for 5.4, and another for 5.10:
- The cut partial erase & 5.4 commits are both needed for 5.4
- Only the 5.10 commit is needed for 5.10.
- The mtd-utils-debug, and mikrotik commits are just for demonstration / testing.

@Leo-PL Yes, your u-boot-env will not be writeable with that config, and only this patch. This patch only exposes the ability for the kernel to use SPI-NOR erase operations for both 4K (where hardware supported) & 64K, rather than only one. If your u-boot-env can be padded (if there is nothing around it) to start and end on an eraseblock boundary, it will work with this. Otherwise, something has to copy the eraseblock, erase it, set the u-boot-env settings, then restore the rest of the eraseblock (like the 411 patch does, but it is not available for 5.10).

<details><summary>My testing on a mikrotik hexs / 760igs</summary>

```
cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00040000 00010000 "RouterBoot"
mtd1: 0000f000 00010000 "bootloader1"
mtd2: 00001000 00010000 "hard_config"
mtd3: 0000f000 00010000 "bootloader2"
mtd4: 00001000 00001000 "soft_config"
mtd5: 00001000 00010000 "bios"
mtd6: 00fc0000 00010000 "firmware"
mtd7: 00310000 00010000 "kernel"
mtd8: 00cb0000 00010000 "rootfs"
mtd9: 00a00000 00010000 "rootfs_data"

echo "ethonce" >  /sys/firmware/mikrotik/soft_config/boot_device

dd if=/dev/mtd0 of=/tmp/mtd0.img

echo -n 'file spi-nor.c +p' > /sys/kernel/debug/dynamic_debug/control
echo -n 'file mtdpart.c +p' > /sys/kernel/debug/dynamic_debug/control
echo -n 'file core.c +p' > /sys/kernel/debug/dynamic_debug/control

echo 1 > /sys/firmware/mikrotik/soft_config/commit

dmesg | tail
[   83.820247] spi-nor spi0.0: at 0x20000, len 4096
[   83.849931] spi-nor spi0.0: to 0x00020000, len 4096


dd if=/dev/mtd0 of=/tmp/mtd0_post.img

cmp -l /tmp/mtd0.img /tmp/mtd0_post.img
131077 165 246
131078  77 117
131079 207  73
131080 324  13
131125   1   3
# only soft CRC32 & booter changed
```
</details>

Cheers

CodeFetch commented:
@Leo-PL Maybe you're looking for kmod-mtd-rw?!

CodeFetch commented:
@john-tho If you reply to your own email the kernel dev's will think someone was assigned already and don't read it. Please try again and without the RFC :grin:...

Leo-PL commented:
> @Leo-PL Maybe you're looking for kmod-mtd-rw?!

Nope, I marked my `uboot-env` partition as RW explicitly.

CodeFetch commented on target/linux/generic/pending-5.10/499-mtd-spi-nor-variable-erase.patch:
```diff
@@ -0,0 +1,204 @@
+-	if (!(child->flags & MTD_NO_ERASE))
++	if (!(child->flags & MTD_NO_ERASE)) {
+ 		wr_alignment = child->erasesize;
++		if (IS_ENABLED(CONFIG_MTD_SPI_NOR_USE_VARIABLE_ERASE) && child->erasesize_minor)
```
The same as above.

john-tho replied:
Moved this `CONFIG` guard to setting the `erasesize_minor` in `spi-nor/core.c`, and took away the `erasesize_minor` test.

CodeFetch commented on target/linux/generic/pending-5.10/499-mtd-spi-nor-variable-erase.patch:
```diff
@@ -0,0 +1,204 @@
+-		child->flags &= ~MTD_WRITEABLE;
+-		printk(KERN_WARNING"mtd: partition \"%s\" doesn't start on an erase/write block boundary -- force read-only\n",
+-			part->name);
++		if (wr_alignment_minor) {
```
Is this `wr_alignment_minor` needed? Why not use `child->erasesize_minor` directly?

john-tho replied:
Matching the existing wr_alignment & erasesize usage.

john-tho replied:
For now, I kept `wr_alignment_minor` so that it is only set `if (!(child->flags & MTD_NO_ERASE))`

CodeFetch commented on target/linux/generic/pending-5.10/499-mtd-spi-nor-variable-erase.patch:
```diff
@@ -0,0 +1,204 @@
++			tmp = mtd_get_master_ofs(child, 0);
++			remainder_minor = do_div(tmp, wr_alignment_minor);
++			if (remainder_minor == 0)
++				child->erasesize = child->erasesize_minor;
```
Why do you set the `erasesize` here to `erasesize_minor`? That doesn't make sense to me...

john-tho replied:
So that if a program uses the erasesize for a partition, to determine the minimum it needs to erase to rewrite a section of the parition, it does not get blocked by the (large) erasesize possibly exceeding the partition bounds.
Example: OpenWrt's `mtd`, in redboot FIS directory rewriting, uses the partition erasesize, but should probably use the partition size: https://github.com/openwrt/openwrt/pull/3271#issuecomment-696422410

CodeFetch replied:
I see... Think this will take a little more time and maybe it makes sense to refactor the code into helper functions as it is not self-explaining and quite chaotic compared to other kernel code.

f00b4r0 replied:
I think this is fine as is, however a quick comment in the code might help make this clearer and save review time upstream.

Still, I wonder if you wouldn't want to adjust `wr_alignment` here: it doesn't make sense from a performance point of view to redo the whole division for the end of the partition if you already know that you are _starting_ on a minor boundary (although this is not in a perf-critical path).

Also correct me if I'm wrong but I don't think you need to fetch the master_ofs again: after the initial division (e.g. 64K EB), all you need to check is whether or not the remainder is a multiple of the minor erase since the minor erase is a multiple of the larger one. And you probably don't need the `remainder_minor` variable either.

All in all, assuming the above is correct, I'd write this like so (which saves a few checks and branches):

```c
tmp = mtd_get_master_ofs(child, 0);
remainder = do_div(tmp, wr_alignment);
if ((child->flags & MTD_WRITEABLE) && remainder) {
	if (wr_alignment_minor) {
		tmp = remainder;
		remainder = do_div(tmp, wr_alignment_minor);
		if (remainder == 0) {
			wr_alignment = wr_alignment_minor;
			/* Some programs check the partition erasize before writing, adjust it */
			child->erasesize = child->erasesize_minor;
		}
	}

	if (remainder != 0) {
		child->flags &= ~MTD_WRITEABLE;
		printk(KERN_WARNING"mtd: partition \"%s\" doesn't start on an erase/write block boundary -- force read-only\n",
			part->name);
	}
}
```

Likewise for the end (without the `wr_alignment` assignment of course).

f00b4r0 replied:
Note that a likely comment from upstream for the child->erasesize assignment may be that the kernel doesn't care about broken programs ;)

john-tho replied:
> a quick comment in the code might help make this clearer

I have not worked out what I want it to say yet… Is this clear:?
Ensure a program which requests an (aligned) mtd_info->erasesize sized erase
will not be denied for the erase overflowing the partition end.

> adjust wr_alignment here: it doesn't make sense from a performance point of view to redo the whole division

Could break the `if (child->flags & MTD_WRITEABLE)` to above the division? One for start, and another for end.


> Also correct me if I'm wrong but I don't think you need to fetch the master_ofs again:

I might introduce another variable for this offset, rather than repeatedly calling it.

> after the initial division (e.g. 64K EB), all you need to check is whether or not the remainder is a multiple of the minor erase since the minor erase is a multiple of the larger one.

Yes, I expect that is the case, but I do know that is it guaranteed.

> And you probably don't need the remainder_minor variable either.

Yes, I like that, thanks.

Cheers

f00b4r0 replied:
> > a quick comment in the code might help make this clearer
> 
> I have not worked out what I want it to say yet… Is this clear:? Ensure a program which requests an (aligned) mtd_info->erasesize sized erase will not be denied for the erase overflowing the partition end.

How about: "Adjust partition erasesize: userspace requesting erasesize-sized erases for minor-aligned partitions will otherwise fail" ?

Still, I suspect this will trigger a NACK upstream, unless there's a very good reason the "program" cannot be fixed :)

> > adjust wr_alignment here: it doesn't make sense from a performance point of view to redo the whole division
> 
> Could break the `if (child->flags & MTD_WRITEABLE)` to above the division? One for start, and another for end.

I'm not sure I follow: if you succcessfully adjust the erasesize at "start", the next `& MTD_WRITEABLE` check is taken and the same computation is done for "end". If you didn't, then `& MTD_WRITEABLE` is not taken and the computation is not performed.

My suggestion was that in the first case, because you adjusted the erasesize at "start", it didn't make sense to first compute the "end" remainder on the larger erasesize which you already know you won't use (although it doesn't hurt). Or am I confused ?

> > Also correct me if I'm wrong but I don't think you need to fetch the master_ofs again:
> 
> I might introduce another variable for this offset, rather than repeatedly calling it.
> 
> > after the initial division (e.g. 64K EB), all you need to check is whether or not the remainder is a multiple of the minor erase since the minor erase is a multiple of the larger one.
> 
> Yes, I expect that is the case, but I do know that is it guaranteed.

Not sure what you mean? By (hardware) design eraseblocks are powers of two. A larger one is thus always a multiple of any smaller one.

> > And you probably don't need the remainder_minor variable either.
> 
> Yes, I like that, thanks.

np, yw :)

HTH

CodeFetch reviewed (changes requested):
I don't grasp what you are doing here. Thus I stopped reviewing... Please use kernel coding style and think about the variables you introduce here.

john-tho commented on target/linux/generic/pending-5.10/499-mtd-spi-nor-variable-erase.patch:
```diff
@@ -0,0 +1,204 @@
+ 		/* Single erase size */
+ 		child->erasesize = master->erasesize;
++		if (master->erasesize_minor)
++			child->erasesize_minor = master->erasesize_minor;
```
I wanted to keep it meaningful: >0 and <erasesize
`BUG_ON(child->erasesize == 0)` steered me away from erasesize_minor 0

john-tho replied:
Did this, as `mtd_info child` is `kzalloc`ed

john-tho commented:
This has had no useful feedback here, in openwrt-devel, or in linux-mtd. Will close.

f00b4r0 commented:
@john-tho I wouldn't close this even if it's not making progress. It's a good reference for why we can't have writeable soft_config on some mikrotik devices, if anything. Cheers

Leo-PL commented:
I second that. The same holds true for cases such as U-boot with writable environment at the last sector of its partition, as I mentioned previously - exactly as implemented in pepe2k/u-boot_mod. So I believe this is still worth pursuing.

dmascord commented:
What is the main blocker for this?  The readability of the code or the functionality ?  This is preventing OpenWRT being useful on Mikrotik devices, I am sure between us all, we can get the code/PR into a state where it can be merged.

CodeFetch commented:
@dmascord The main blocker is that this belongs upstream, but it was ignored by the kernel devs. I think the current state in this PR adds complexity which would need code refactoring to make it better understandable, but that is only my subjective perception.

john-tho committed:
kernel: cut broken SPI_NOR 4K eraseblock LIMIT patch

Since 4e0c54bc5bc8 ("kernel: add support for kernel 5.4"),
the spi-nor limit 4k erasesize to spi-nor chips below a configured size
patch has not functioned as intended.

For uniform erasesize SPI-NOR devices, both
nor->erase_opcode & mtd->erasesize are used in erase operations.
These are set before, and not modified by, this
CONFIG_MTD_SPI_NOR_USE_4K_SECTORS_LIMIT patch.
Thus, an SPI-NOR device with CONFIG_MTD_SPI_NOR_USE_4K_SECTORS will
always use 4k erasesize (where the device supports it).

If this patch was fixed to function as intended, there would be
cases where devices change from a 4K to a 64K erasesize.

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>

john-tho commented on target/linux/generic/pending-5.10/499-mtd-spi-nor-variable-erase.patch:
```diff
@@ -0,0 +1,204 @@
+ 		if (i > 0)
+ 			i--;
+ 
+-		/* Pick biggest erasesize */
```
I took out the code to set erasesize_minor for a variable erase-regions type SPI-NOR device. I do not have one to test, and cannot find one easily.

john-tho commented:
I have update the variable erase patch a little.

OpenWrt changes:
- Only apply to 5.10, so no need to deal with incomplete replacement for pending-5.4 411 & 412 (they allow any erase, this allows minor-aligned erases)
- Add the introduced kernel config symbol unset in generic kconfig
- Cut the broken (it does nothing for uniform erase-regions SPI-NOR chips) [pending-5.10 470 4K limit patch](https://github.com/openwrt/openwrt/blob/master/target/linux/generic/pending-5.10/470-mtd-spi-nor-support-limiting-4K-sectors-support-base.patch). It is also broken in 5.4, but not touching 5.4.

Patch changes:
- add sysfs `erasesize_minor`
- guard setting `erasesize_minor` in `spi-nor/core.c` with a Kconfig, rather than (the SPI-NOR specific Kconfig) guarding in `mtdpard.c`
- cut finding minor erasesize for variable erase regions device. Do not have a relevant SPI-NOR chip to test

john-tho committed:
ipq40xx: avoid parsing well defined NOR partitions

These partitions have fixed offsets on all mikrotik ipq40xx devices
No need to parse at startup to find them

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>

hauke commented:
Some of this was added in commit 8441a622eca236c5ed19545c7b11129466c1d8cb. 
It would still be nice to get this integrated into upstream Linux, if you do not get an answer in one month you can resend the patch again.

hauke commented on target/linux/generic/pending-5.10/470-mtd-spi-nor-support-limiting-4K-sectors-support-base.patch:
```diff
@@ -1,71 +0,0 @@
- 	 */
- 	erase_mask = 0;
- 	i = 0;
-+#ifdef CONFIG_MTD_SPI_NOR_USE_4K_SECTORS
```
I do not understand why this is not working based on the commit message.
If the flash has `SECT_4K` or `SECT_4K_PMC` set we add an erase operation for 4k in addition to the normal erase operation.

In `spi_nor_select_erase()` the call to `spi_nor_has_uniform_erase()` will return true and if `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS` is set we would select the 4k erase operation and disable the other.

john-tho replied:
> we would select the 4k erase operation and disable the other.

`CONFIG_MTD_SPI_NOR_USE_4K_SECTORS` already does this.

As is, the `LIMIT` does not work, and makes this 470 patch do nothing (only the depended on  `USE_4K_SECTORS` matters). 

For uniform erase regions devices:
`spi_nor_erase_sector` does not use the `erase_map`. It uses [`nor->erase_opcode`](https://elixir.bootlin.com/linux/v5.10.105/C/ident/erase_opcode), which was set in `spi_nor_select_erase`, called by `spi_nor_default_setup` near the start of `spi_nor_info_init_params`, before these 470 changes are run, or possibly in one of the [later `init_params` functions](https://elixir.bootlin.com/linux/v5.10.105/source/drivers/mtd/spi-nor/core.c#L2844).

Might fix it by preventing changing the [`wanted_size`](https://elixir.bootlin.com/linux/v5.10.105/source/drivers/mtd/spi-nor/core.c#L2577) if device size > LIMIT.

f00b4r0 commented:
@john-tho please consider the changes I suggested, they may help mainlining, if only by reducing the patch size.

john-tho commented:
Thank you hauke and f00,
I did start to implement some of your suggestions, but got busy before I finished testing and pushing them. I will get back to this.
Would have been nice to get input into/from 8441a622eca236c5ed19545c7b11129466c1d8cb, rather than someone grab the patch from linux-mtd, and resubmit it to openwrt-devel. They did enable it for _many_ devices in ramips/mt7621, so it has had significantly more basic testing than beforehand…

f00b4r0 commented:
@john-tho I have rebased your work on current master and implemented the changes I suggested above; it's all available here:
https://github.com/f00b4r0/openwrt/tree/mtd64k
The branch includes the reversal of 8441a622eca236c5ed19545c7b11129466c1d8cb which IMHO should never have been merged without your input.
It only deals with 5.10 for now.

This code removes the "workaround" for buggy userspace (the `child->erasesize = child->erasesize_minor;` part). I've **only done some happy testing** with `soft_config` on ath79 devices: it works as intended. routerboot2 does not get corrupted when writing to soft_config and writes are correctly handled.

Of course that means that only the start adjustment path has been tested, but in this version of the patch there is no dependency between the start and end adjustments, so I expect the end adjustment to work identically well.

I think this is more likely to make the patch acceptable upstream, but I dump below a separate incremental patch that restores the workaround if needed.

I haven't touched the commit messages at this stage (save for the addition of the v2->v3 changelog).

```
--- a/drivers/mtd/mtdpart.c
+++ b/drivers/mtd/mtdpart.c
@@ -182,6 +182,11 @@ static struct mtd_info *allocate_partition(struct mtd_info *parent,
 		if (wr_alignment_minor) {
 			tmp = remainder;
 			remainder = do_div(tmp, wr_alignment_minor);
+			if (!remainder) {
+				/* Adjust partition erasesize: userspace requesting erasesize-sized
+				 * erases for minor-aligned partitions will otherwise fail */
+				child->erasesize = child->erasesize_minor;
+			}
 		}
 
 		if (remainder) {
@@ -197,6 +203,8 @@ static struct mtd_info *allocate_partition(struct mtd_info *parent,
 		if (wr_alignment_minor) {
 			tmp = remainder;
 			remainder = do_div(tmp, wr_alignment_minor);
+			if (!remainder)
+				child->erasesize = child->erasesize_minor;
 		}
 
 		if (remainder) {
```

Let me know if you want me to submit this upstream myself (if you don't have time to deal with this anymore).

HTH

john-tho committed:
kernel: fix variable erasesizes patch

Update this pending patch to remove the untested (variable eraseregions)
section, alongside simplifying the patch.

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>

john-tho committed:
ath79: mikrotik: add rw soft_config to extra devices

Linux MTD requires the parent partition be writable for a child
partition to be allowed write permission.

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>

john-tho commented:
Pushed a refresh + update of this patch.
Have done some testing.
Will send this as a v2 to linux-mtd

f00b4r0 commented:
@xback would you consider committing this much improved version to replace 8441a62 which was pulled from here?

xback commented:
> @xback would you consider committing this much improved version to replace [8441a62](https://github.com/openwrt/openwrt/commit/8441a622eca236c5ed19545c7b11129466c1d8cb) which was pulled from here?

If it has your blessing I'll be more than happy to give it a spin :+1:

f00b4r0 commented:
> > @xback would you consider committing this much improved version to replace [8441a62](https://github.com/openwrt/openwrt/commit/8441a622eca236c5ed19545c7b11129466c1d8cb) which was pulled from here?
> 
> If it has your blessing I'll be more than happy to give it a spin 👍

It certainly does. The current version is much cleaner than what was merged in openwrt :)

Leo-PL commented:
@xback, @f00b4r0, @john-tho: A tiny warning: it is possible, that certain devices may now lose configuration over `sysupgrade`, because padding before configuration tarball after the end of rootfs will change during the upgrade. Perhaps we should update `sysupgrade` to look for configuration to restore in next 4k AND next 64k-aligned block after upgrade, to avoid requiring manual restore. This will be the case, for example with `rt305x` subtarget of `ramips`. I'll give a spin of this PR on my MF283+, it'll see great performance improvement from that, but may lose the configuration during update. I'll keep you posted.

f00b4r0 commented:
> > > @xback would you consider committing this much improved version to replace [8441a62](https://github.com/openwrt/openwrt/commit/8441a622eca236c5ed19545c7b11129466c1d8cb) which was pulled from here?
> > 
> > 
> > If it has your blessing I'll be more than happy to give it a spin 👍
> 
> It certainly does. The current version is much cleaner than what was merged in openwrt :)

OK just a caveat: I did not notice 3a7696d and 34d3016
I think these are completely orthogonal to this PR and should be in a separate PR. I think the other 4 patches are OK. Specifically, bfb2f7d is what I'd like to see merged ASAP as it fixes the code that was previously merged in OpenWRT.

f00b4r0 commented:
> @xback, @f00b4r0, @john-tho: A tiny warning: it is possible, that certain devices may now lose configuration over `sysupgrade`, because padding before configuration tarball after the end of rootfs will change during the upgrade. Perhaps we should update `sysupgrade` to look for configuration to restore in next 4k AND next 64k-aligned block after upgrade, to avoid requiring manual restore. This will be the case, for example with `rt305x` subtarget of `ramips`. I'll give a spin of this PR on my MF283+, it'll see great performance improvement from that, but may lose the configuration during update. I'll keep you posted.

I don't see how what you describe could happen. This code (bfb2f7d) does not introduce new functionality, it merely fixes a previous commit that was grabbed from a WIP version of this PR, which was merged in 8441a622eca236c5ed19545c7b11129466c1d8cb

More importantly, the build recipes are unchanged and the erase size reported to the OS is also unchanged (it will still be 64K). So I think the scenario you describe is impossible.

f00b4r0 commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mikrotik_nor.dtsi:
```diff
@@ -23,25 +23,47 @@
+					label = "soft_config";
+					reg = <0x3d000 0x1000>;
+				};
+			};
```
I think this change is completely orthogonal to this PR and shouldn't be here. I also believe this particular change brings no benefit and will likely need to be reverted when devices which "break the rule" appear (as was the case with ath79, where for a while we thought that "all" devices followed the same partitioning, only to eventually be proven wrong, thus prompting for the implementation of the parser).

f00b4r0 reviewed (changes requested):
Please split these two patches out of this PR:
https://github.com/openwrt/openwrt/pull/3271/commits/3a7696d4617cf974c09e189e68f7b17ee37c257d
https://github.com/openwrt/openwrt/pull/3271/commits/34d301613264722b74eca4bf43121773503a9da0

Leo-PL commented:
Acked, makes sense. But to get better performance, I'd eventually have to disable `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS`, and then it will hit me. My reason for doing so is the fact, that performance of 16MB flash with 4k erase size is abysmal, and `sysupgrade` takes good 8 minutes in this case.

Leo-PL commented:
I figured out how to sysupgrade when changing global MTD erase size from 4k to 64k without losing configuration. It can be done by setting `BLOCKSIZE := 64k` at the same time as disabling `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS`. The new image will be aligned to 64k boundary, so JFFS2 filesystem with `sysupgrade.tgz` will end up in the right place for new alignment after reboot.
It is not possible the other way around, unless `mtd` is modified , but there is no point in doing so. 

Now this would be quite a big perk for ramips, especially for mt7620, rt305x, and rt3883, as some of them have partitions smaller than 64k. though most of them need not be writable. 
OTOH, I'm not sure if it's possible to change that globally without splitting those subtargets, as they contain a few tiny devices having only 4MB of Flash, all disabled by now.

f00b4r0 commented:
@xback @john-tho for your convenience I have rebased this PR on current master and removed the extra 2 commits which IMO should be part of a separate PR.

I have also split [bfb2f7d](https://github.com/openwrt/openwrt/pull/3271/commits/bfb2f7dc05848c69e5597e98dbcc14f234b2582a) into two separate commits (with no change at all): the actual commit and the patch refreshes, which should make it easier to rebase later on (I had a conflict when I rebased).

This is built tested and available here: https://github.com/f00b4r0/openwrt/commits/mtd64k

xback commented:
Pushed rebased version from @f00b4r0 to master.
Thanks for the patches!

f00b4r0 commented:
@john-tho is it OK if I submit this upstream? I think we should at least try to get this upstreamed.

john-tho commented:
> @john-tho is it OK if I submit this upstream? I think we should at least try to get this upstreamed.

Sure.

f00b4r0 commented:
> > @john-tho is it OK if I submit this upstream? I think we should at least try to get this upstreamed.
> 
> Sure.

OK, I'll get on it! I'll probably want to reword the commit message a bit, I'll run it by you before I send :)
